File: src/core/player.rs
use crate::core::stepfile::StepsType;
use bevy::prelude::*;
use std::ops::{Index, IndexMut};
use strum::{EnumIter, IntoStaticStr};

/// One of the two player slots the machine offers. Every player-scoped
/// piece of state — settings, high scores, key bindings, a play session's
/// stage — is keyed by this. `Default` exists only because BSN's
/// `FromTemplate` derive demands it of component fields; no game logic
/// falls back to it.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, EnumIter, IntoStaticStr)]
pub enum PlayerId {
    #[default]
    P1,
    P2,
}

impl PlayerId {
    pub fn label(self) -> &'static str {
        self.into()
    }
}

/// One value per player slot, for state that always exists for both
/// players regardless of how many are active.
#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct PerPlayer<T> {
    pub p1: T,
    pub p2: T,
}

impl<T> Index<PlayerId> for PerPlayer<T> {
    type Output = T;

    fn index(&self, player: PlayerId) -> &T {
        match player {
            PlayerId::P1 => &self.p1,
            PlayerId::P2 => &self.p2,
        }
    }
}

impl<T> IndexMut<PlayerId> for PerPlayer<T> {
    fn index_mut(&mut self, player: PlayerId) -> &mut T {
        match player {
            PlayerId::P1 => &mut self.p1,
            PlayerId::P2 => &mut self.p2,
        }
    }
}

/// How the machine is being played: who is on it and which charts they
/// step. Selected on the mode select scene and read by every scene after
/// it.
#[derive(Resource, Debug, Clone, Copy, Default, PartialEq, EnumIter, IntoStaticStr)]
pub enum PlayMode {
    #[default]
    Singles,
    Doubles,
    Versus,
}

impl PlayMode {
    pub fn label(self) -> &'static str {
        self.into()
    }

    /// The chart type this mode plays.
    pub fn steps_type(self) -> StepsType {
        match self {
            PlayMode::Singles | PlayMode::Versus => StepsType::DanceSingle,
            PlayMode::Doubles => StepsType::DanceDouble,
        }
    }

    /// The active player slots: P1 alone plays singles and doubles (the
    /// doubles chart spans both pads), versus fields both.
    pub fn players(self) -> &'static [PlayerId] {
        match self {
            PlayMode::Singles | PlayMode::Doubles => &[PlayerId::P1],
            PlayMode::Versus => &[PlayerId::P1, PlayerId::P2],
        }
    }
}

File: core/Player.cs
namespace Rhythm.Core;

/// <summary>
/// One of the two player slots the machine offers. Every player-scoped
/// piece of state — settings, high scores, key bindings, a play session's
/// stage — is keyed by this.
/// </summary>
public enum PlayerId
{
    P1,
    P2,
}

/// <summary>
/// One value per player slot, for state that always exists for both
/// players regardless of how many are active.
/// </summary>
public struct PerPlayer<T>(T p1, T p2)
{
    public T P1 = p1;
    public T P2 = p2;

    public static PerPlayer<T> Uniform(T value) => new(value, value);

    public T this[PlayerId player]
    {
        readonly get => player == PlayerId.P1 ? P1 : P2;
        set
        {
            if (player == PlayerId.P1)
            {
                P1 = value;
            }
            else
            {
                P2 = value;
            }
        }
    }
}

/// <summary>
/// How the machine is being played: who is on it and which charts they
/// step. Selected on the mode select scene (kept on the <c>Game</c> root)
/// and read by every scene after it.
/// </summary>
public enum PlayMode
{
    Singles,
    Doubles,
    Versus,
}

public static class PlayModeExtensions
{
    /// <summary>The chart type this mode plays.</summary>
    public static StepsType StepsType(this PlayMode mode) =>
        mode == PlayMode.Doubles ? Rhythm.Core.StepsType.DanceDouble : Rhythm.Core.StepsType.DanceSingle;

    /// <summary>
    /// The active player slots: P1 alone plays singles and doubles (the
    /// doubles chart spans both pads), versus fields both.
    /// </summary>
    public static IReadOnlyList<PlayerId> Players(this PlayMode mode) =>
        mode == PlayMode.Versus ? TwoPlayers : OnePlayer;

    private static readonly PlayerId[] OnePlayer = [PlayerId.P1];
    private static readonly PlayerId[] TwoPlayers = [PlayerId.P1, PlayerId.P2];
}

/// <summary>
/// A step panel's direction, in the fixed Left/Down/Up/Right column order
/// every 4-panel pad uses.
/// </summary>
public enum StepDirection
{
    Left,
    Down,
    Up,
    Right,
}

public static class StepDirectionExtensions
{
    /// <summary>The direction of a pad-local column (<c>0..4</c>).</summary>
    public static StepDirection OfColumn(int column) =>
        column switch
        {
            0 => StepDirection.Left,
            1 => StepDirection.Down,
            2 => StepDirection.Up,
            _ => StepDirection.Right,
        };

    /// <summary>The pad-local column of this direction — <see cref="OfColumn"/>'s inverse.</summary>
    public static int Column(this StepDirection direction) => (int)direction;
}
